dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,472 @@
+            for idx, input_name in enumerate(self.dali_input_names):
+                self.pipe.feed_input(input_name, inputs[idx])
+            self.pipe._run_once()
+            pipe_outputs = self.pipe.outputs()
+            torch_outputs = tuple(
+                [
+                    to_torch_tensor(out.as_tensor(), not self.pipe.exec_dynamic)
+                    for out in pipe_outputs
+                ]
+            )
+            self.dali_output_q.put((batch_id, torch_outputs))
+            torch.cuda.nvtx.range_pop()
+
+    def start_thread(self):
+        """
+        Starts the DALI pipeline thread
+        """
+        if self.thread is not None:
+            return
+        self.thread = threading.Thread(target=DALIServer.thread_fn, args=(self,))
+        self.thread_stop_event = threading.Event()
+        self.thread.start()
+
+    def stop_thread(self):
+        """
+        Stops the DALI pipeline thread
+        """
+        if self.thread_stop_event is None:
+            return
+        self.thread_stop_event.set()
+        self.thread.join()
+        self.thread = None
+        self.thread_stop_event = None
+
+    def __enter__(self):
+        self.start_thread()
+        return self
+
+    def __exit__(self, exc_type, exc_value, tb):
+        self.stop_thread()
+        if exc_type is not None:
+            warnings.warn(f"An exception occurred: {exc_value}", category=UserWarning)
+        return False  # Return False to propagate exceptions
+
+
+def _collate_dali_processed_sample_ref_fn(pipe_out, *, collate_fn_map=None):
+    """
+    Special collate function that schedules a DALI iteration for execution
+    """
+    assert len(pipe_out) > 0
+    first_elem = pipe_out[0]
+    inputs = [[] for _ in range(len(first_elem.inputs))]
+    proxy = first_elem.proxy
+    for elem in pipe_out:
+        assert proxy == elem.proxy
+        for idx, input_ref in enumerate(elem.inputs):
+            inputs[idx].append(input_ref)
+    ret = proxy._schedule_batch(inputs)
+    return ret
+
+
+# In-place modify `default_collate_fn_map` to handle DALIProcessedSampleRef
+default_collate_fn_map.update({DALIProcessedSampleRef: _collate_dali_processed_sample_ref_fn})
+
+
+class DataLoader(torchdata.dataloader.DataLoader):
+    """
+    DALI data loader to be used in the main loop, which runs the DALI pipeline doing the
+    processing asynchronously with regards to the training.
+    """
+
+    class _Iter(torchdata.dataloader._MultiProcessingDataLoaderIter):
+        """
+        Data loader iterator used by the DALI proxy data loader
+        """
+
+        def __init__(self, loader):
+            super().__init__(loader)
+            self.loader = loader
+
+        def _next_data(self):
+            data = super()._next_data()
+            if not hasattr(data, "__iter__"):
+                warnings.warn(
+                    "Warning: Non iterable returned from dataloader. Please "
```

Comment:
Nitpick: I believe we no longer need "Warning: " in the message, `warn` will take care of that